source/common/http/http1/codec_impl.cc

```diff
@@ -970,20 +971,23 @@ void ServerConnectionImpl::onResetStream(StreamResetReason reason) {
 }
 
 void ServerConnectionImpl::sendProtocolError(absl::string_view details) {
-  if (active_request_.has_value()) {
-    active_request_.value().response_encoder_.setDetails(details);
-  }
-  // We do this here because we may get a protocol error before we have a logical stream. Higher
-  // layers can only operate on streams, so there is no coherent way to allow them to send an error
-  // "out of band." On one hand this is kind of a hack but on the other hand it normalizes HTTP/1.1
-  // to look more like HTTP/2 to higher layers.
-  if (!active_request_.has_value() ||
-      !active_request_.value().response_encoder_.startedResponse()) {
-    Buffer::OwnedImpl bad_request_response(
-        absl::StrCat("HTTP/1.1 ", error_code_, " ", CodeUtility::toString(error_code_),
-                     "\r\ncontent-length: 0\r\nconnection: close\r\n\r\n"));
-
-    connection_.write(bad_request_response, false);
+  // We do this here because we may get a protocol error before we have a logical stream.
+  if (!active_request_.has_value()) {
+    // TODO(alyssawilk) integration test the early response error path in overload.
+    onMessageBeginBase();
+  }
+  ASSERT(active_request_.has_value());
+
+  active_request_.value().response_encoder_.setDetails(details);
+  if (!active_request_.value().response_encoder_.startedResponse()) {
+    // TODO(alyssawilk) best-effort set these.
+    bool is_grpc_request = false;
+    bool is_head_request = false;
+    active_request_->request_decoder_->sendLocalReply(
```

Comment:
@yanavlasov Correct me if I'm wrong, this is my read on this

`sendLocalReply` will `encodeHeaders`, which currently may throw and will be moved to RELEASE_ASSERTs and then proper clean-up with the exception removal process. Before this PR, these were outside a dispatching callstack and are uncaught.

Here, we are in a dispatching callstack. We would need to do two things to mesh well with this PR. (1) add return status for sendProtocolError/sendLocalReply and handle the way we've been handling error statuses while dispatching and (2) wrap `RELEASE_ASSERT`s in things like `encodeHeaders` that add for future clean-up (https://github.com/envoyproxy/envoy/pull/11575/files#r443142918) with a test for `isNotDispatching()`.

Reply:
Have we started work for the HTTP/1 coded?  I saw Yan's PR for the HTTP/2 codec but I wasn't sure what the plans were for HTTP/1.  Is he doing both? If so I assume I can just land as it'll be at least a few weeks until HTTP/2 is done and HTTP/1 changes matter

Reply:
I made an H/1 PR a while ago (it was reverted, it is not merged) https://github.com/envoyproxy/envoy/pull/11101
There is one exception still there in that PR in encodeHeaders (non-dispatching), which I will handle however Yan decides to settle the RELEASE_ASSERT todo

Reply:
I think going ahead makes sense, I can implement the status return values for sendProtocolError  I mentioned above if this PR is added.

Reply:
I did not think that H/1 codec's `encodeHeader()` was throwing. If it did not throw before then there should be no problem with this change.

Reply:
Oh, maybe it will only call the responseencoder's? If so, that makes this easier. https://github.com/envoyproxy/envoy/blob/77efb4799d906400b97b43d6c2ed0b5d8369b2fc/source/common/http/http1/codec_impl.cc#L390